File: wrappers/java/src/test/java/org/hyperledger/indy/sdk/ledger/GetTxnRequestTest.java
package org.hyperledger.indy.sdk.ledger;

import static org.junit.Assert.assertNotNull;
import static org.junit.Assert.assertTrue;

import org.hyperledger.indy.sdk.IndyIntegrationTestWithPoolAndSingleWallet;
import org.hyperledger.indy.sdk.utils.PoolUtils;
import org.json.JSONObject;
import org.junit.Test;
import org.hyperledger.indy.sdk.JsonObjectSimilar;

public class GetTxnRequestTest extends IndyIntegrationTestWithPoolAndSingleWallet {

	@Test
	public void testBuildGetTxnRequestWorks() throws Exception {
		int seq_no = 1;

		JSONObject expectedResult = new JSONObject()
				.put("identifier", DID)
				.put("operation",
						new JSONObject()
								.put("type", "3")
								.put("ledgerId", 1)
								.put("data", seq_no)
				);

		String getTxnRequest = Ledger.buildGetTxnRequest(DID, null, seq_no).get();
		assert (new JSONObject(getTxnRequest).toMap().entrySet()
				.containsAll(
						expectedResult.toMap().entrySet()));
	}

	@Test
	public void testBuildGetTxnRequestWorksForLedgerType() throws Exception {
		int seq_no = 1;

		JSONObject expectedResult = new JSONObject()
				.put("identifier", DID)
				.put("operation",
						new JSONObject()
								.put("type", "3")
								.put("ledgerId", 0)
								.put("data", seq_no)
				);

		String getTxnRequest = Ledger.buildGetTxnRequest(DID, "POOL", seq_no).get();
		assert (new JSONObject(getTxnRequest).toMap().entrySet()
				.containsAll(
						expectedResult.toMap().entrySet()));
	}

	@Test(timeout = PoolUtils.TEST_TIMEOUT_FOR_REQUEST_ENSURE)
	public void testGetTxnRequestWorks() throws Exception {
		String did = createStoreAndPublishDidFromTrustee();

		String schemaRequest = Ledger.buildSchemaRequest(did, SCHEMA_DATA).get();
		String schemaResponse = Ledger.signAndSubmitRequest(pool, wallet, did, schemaRequest).get();

		JSONObject schemaResponseObj = new JSONObject(schemaResponse);
		int seqNo = schemaResponseObj.getJSONObject("result").getJSONObject("txnMetadata").getInt("seqNo");

		String getTxnRequest = Ledger.buildGetTxnRequest(did, null, seqNo).get();
		String expectedData = "{\"name\":\"gvt\",\"version\":\"1.0\",\"attr_names\": [\"name\"]}";

		String getTxnResponse = PoolUtils.ensurePreviousRequestApplied(pool, getTxnRequest, response -> {
			JSONObject getTxnResponseObj = new JSONObject(response);
			JSONObject schemaTransactionObj =
					getTxnResponseObj.getJSONObject("result").getJSONObject("data").getJSONObject("txn").getJSONObject("data").getJSONObject("data");

			return JsonObjectSimilar.similar(new JSONObject(expectedData), schemaTransactionObj);
		});
		assertNotNull(getTxnResponse);
	}
}

File: wrappers/dotnet/indy-sdk-dotnet-test/LedgerTests/GetTxnRequestTest.cs
using Hyperledger.Indy.LedgerApi;
using Microsoft.VisualStudio.TestTools.UnitTesting;
using Newtonsoft.Json.Linq;
using System.Threading.Tasks;

namespace Hyperledger.Indy.Test.LedgerTests
{
    [TestClass]
    public class GetTxnRequestTest : IndyIntegrationTestWithPoolAndSingleWallet
    {        
        [TestMethod]
        public async Task TestBuildGetTxnRequestWorks()
        {
            var seq_no = 1;

            var expectedResult = string.Format("\"identifier\":\"{0}\"," +
                    "\"operation\":{{" +
                    "\"type\":\"3\"," +
                    "\"data\":{1}," +
                    "\"ledgerId\":1" +
                    "}}", DID, seq_no);

            var getTxnRequest = await Ledger.BuildGetTxnRequestAsync(DID, null, seq_no);

            Assert.IsTrue(getTxnRequest.Replace("\\", "").Contains(expectedResult));
        }

        [TestMethod]
        public async Task TestBuildGetTxnRequestWorksForLedgerType()
        {
            var seq_no = 1;

            var expectedResult = string.Format("\"identifier\":\"{0}\"," +
                    "\"operation\":{{" +
                    "\"type\":\"3\"," +
                    "\"data\":{1}," +
                    "\"ledgerId\":0" +
                    "}}", DID, seq_no);

            var getTxnRequest = await Ledger.BuildGetTxnRequestAsync(DID, "POOL", seq_no);

            Assert.IsTrue(getTxnRequest.Replace("\\", "").Contains(expectedResult));
        }

        [TestMethod] //This test fails here and in the Java version.
        public async Task TestGetTxnRequestWorks()
        {
            var did = await CreateStoreAndPublishDidFromTrusteeAsync();

            var schemaRequest = await Ledger.BuildSchemaRequestAsync(did, SCHEMA_DATA);
            var schemaResponse = await Ledger.SignAndSubmitRequestAsync(pool, wallet, did, schemaRequest);

            var schemaResponseObj = JObject.Parse(schemaResponse);

            var seqNo = schemaResponseObj["result"]["txnMetadata"].Value<int>("seqNo");

            var getTxnRequest = await Ledger.BuildGetTxnRequestAsync(did, null, seqNo);
            var expectedData = "{\"name\":\"gvt\",\"version\":\"1.0\",\"attr_names\": [\"name\"]}";

            var getTxnResponse = await PoolUtils.EnsurePreviousRequestAppliedAsync(pool, getTxnRequest, response => {
                var getTxnResponseObj = JObject.Parse(response);
                var schemaTransactionObj = getTxnResponseObj["result"]["data"]["txn"]["data"]["data"];

                return JValue.DeepEquals(JObject.Parse(expectedData), schemaTransactionObj);
            });

            Assert.IsNotNull(getTxnResponse);
        }

        [TestMethod] 
        public async Task TestGetTxnRequestWorksForInvalidSeqNo()
        {
            var did = await CreateStoreAndPublishDidFromTrusteeAsync();

            var schemaRequest = await Ledger.BuildSchemaRequestAsync(did, SCHEMA_DATA);
            var schemaResponse = await Ledger.SignAndSubmitRequestAsync(pool, wallet, did, schemaRequest);

            var schemaResponseObj = JObject.Parse(schemaResponse);

            var seqNo = (int)schemaResponseObj["result"]["txnMetadata"]["seqNo"] + 1;

            var getTxnRequest = await Ledger.BuildGetTxnRequestAsync(did, null, seqNo);
            var getTxnResponse = await Ledger.SubmitRequestAsync(pool, getTxnRequest);

            var getTxnResponseObj = JObject.Parse(getTxnResponse);

            Assert.IsTrue(getTxnResponseObj["result"]["data"] != null);
        }
    }
}
